### PR TITLE
drivers: sensor: lsm6dsl: fix precision on gyro and acc measurements

### PR DESCRIPTION
**Pull request by Alessio-Parmeggiani**
Following #63980, I suggest a fix to allow for more than 3 decimal digits on gyroscope and accelerations measurements from the LSM6DSL sensor.

### CODE REVIEW

**Comment by arbrauns** [comment]
This should instead use `sensor_value_from_double()`, which also includes bounds checking.

File: drivers/sensor/lsm6dsl/lsm6dsl.c [hunk 1]
```diff
@@ -478,7 +478,7 @@ static inline void lsm6dsl_accel_convert(struct sensor_value *val, int raw_val,
 	/* Convert to m/s^2 */
 	dval = (double)(raw_val) * (double)sensitivity * SENSOR_G_DOUBLE / 1000;
 	val->val1 = (int32_t)dval;
-	val->val2 = (((int32_t)(dval * 1000)) % 1000) * 1000;
+	val->val2= (((int32_t)(dval * 1000000)) % 1000000);
```
Comment (avisconti) [inline drivers/sensor/lsm6dsl/lsm6dsl.c:482]: Please check the spacing as indicated. 
Moreover I'm confused... How is it possible that was working before? The line seems three degrees of order smaller than before. Anyway, now seems correct to me. I would suggest to get rid of SENSOR_G_DOUBLE and use the sensor_g_to_ms2() API.

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/lsm6dsl/lsm6dsl.c (hunk 1 above)
Comment (arbrauns) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:482]: > How is it possible that was working before?

If I got this right, the previous code produced values only accurate up to a thousandth of a `m/s²`, i.e. the three least significant decimal digits in `val2` were always zero. With the new code, it's accurate up to a millionth of a `m/s²` (though it's of course questionable whether the sensor can actually provide anything close to that reliably).

Comment (avisconti) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:482]: @Alessio-Parmeggiani can you squash the two commits and make use of the APIs mentioned above?

Comment (Alessio-Parmeggiani) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:482]: @avisconti I added a commit where I use the suggested API. By using this new method there is no need for the original conversion I proposed in the first commit.

Comment (avisconti) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:482]: Can you possibly squash the two commits, or remove the first one. I would like to see only one commit in the PR

File: drivers/sensor/lsm6dsl/lsm6dsl.c [hunk 2]
```diff
@@ -476,9 +476,9 @@ static inline void lsm6dsl_accel_convert(struct sensor_value *val, int raw_val,
 
 	/* Sensitivity is exposed in mg/LSB */
 	/* Convert to m/s^2 */
-	dval = (double)(raw_val) * (double)sensitivity * SENSOR_G_DOUBLE / 1000;
-	val->val1 = (int32_t)dval;
-	val->val2 = (((int32_t)(dval * 1000)) % 1000) * 1000;
+	dval = (double)(raw_val) * (double)sensitivity / 1000; 
+	int32_t dval_ug = dval * 1000000; //value converted in micro g
+	return sensor_ug_to_ms2(dval_ug, val); //value in m/s^2
```
Comment (avisconti) [inline drivers/sensor/lsm6dsl/lsm6dsl.c:481]: it's a void function, so no need to return a value.

File: drivers/sensor/lsm6dsl/lsm6dsl.c [hunk 3]
```diff
@@ -526,9 +526,9 @@ static inline void lsm6dsl_gyro_convert(struct sensor_value *val, int raw_val,
 
 	/* Sensitivity is exposed in mdps/LSB */
 	/* Convert to rad/s */
-	dval = (double)(raw_val * (double)sensitivity * SENSOR_DEG2RAD_DOUBLE / 1000);
-	val->val1 = (int32_t)dval;
-	val->val2 = (((int32_t)(dval * 1000)) % 1000) * 1000;
+	dval = (double)(raw_val) * (double)sensitivity / 1000; //value in degrees
+	int32_t dval_10ud = dval * 100000; //value in 10 micro degrees
+	return sensor_10udegrees_to_rad(dval_10ud, val); //value in radians
```
Comment (avisconti) [inline drivers/sensor/lsm6dsl/lsm6dsl.c:531]: same here

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/lsm6dsl/lsm6dsl.c (hunk 1 above)
Comment (Alessio-Parmeggiani) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:482]: @avisconti Done!

File: drivers/sensor/lsm6dsl/lsm6dsl.c [hunk 4]
```diff
@@ -476,9 +476,9 @@ static inline void lsm6dsl_accel_convert(struct sensor_value *val, int raw_val,
 
 	/* Sensitivity is exposed in mg/LSB */
 	/* Convert to m/s^2 */
-	dval = (double)(raw_val) * (double)sensitivity * SENSOR_G_DOUBLE / 1000;
-	val->val1 = (int32_t)dval;
-	val->val2 = (((int32_t)(dval * 1000)) % 1000) * 1000;
+	dval = (double)(raw_val) * (double)sensitivity / 1000; 
+	int32_t dval_ug = dval * 1000000; //value converted in micro g
```
Comment (avisconti) [inline drivers/sensor/lsm6dsl/lsm6dsl.c:480]: Sorry if I noticed only now (my bad), but I'm not sure it is allowed to declare a variable in the middle of a body function. I don't remember any specific rule, but I never noticed it. Maybe @fabiobaltieri or @MaureenHelm may confirm it or not.

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/lsm6dsl/lsm6dsl.c (hunk 4 above)
Comment (avisconti) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:480]: Moreover, if you have already a  value espressed in g, maybe you can directly use the sensor_g_to_ms2.
Alternatively we can change the sensitivity in .h and express it as 61 ug/LSB (you can take a look to #71223 that I just wrote looking at this work. I left lsm6dsl out to let you finish it)

**Comment by avisconti** [comment]
@Alessio-Parmeggiani can you pls rebase this commit? We are very close to finish it. Thanks

File: drivers/sensor/lsm6dsl/lsm6dsl.c (hunk 4 above)
Comment (Alessio-Parmeggiani) [reply drivers/sensor/lsm6dsl/lsm6dsl.c:480]: @avisconti Sorry i missed your reply.

I don't understand your comment about the variable in the middle of a body function, do you mean that the variable should be declared at the top, like dval, or that everything should be done without using additional variables?

I'll look at your pull request and add the same modifications to lsm6dsl

**Comment by Alessio-Parmeggiani** [comment]
> @Alessio-Parmeggiani can you pls rebase this commit? We are very close to finish it. Thanks

Sorry, what do you mean? I am not an expert in this kind of git stuff

**Comment by avisconti** [comment]
> > @Alessio-Parmeggiani can you pls rebase this commit? We are very close to finish it. Thanks
> 
> Sorry, what do you mean? I am not an expert in this kind of git stuff

I see now that you are using your main branch. You may want to create a new branch for this work from main (e.g. ```git checkout -b lsm6dsl-fix```). Then come back to main and ```git pull``` all the work. Then switch back to your work branch and rebase it interactively on main (i.e. ```git rebase -i main```). Please take a look at documentation if this is new stuff. But it is something required for a zephyr contributor.

**Comment by avisconti** [comment]
> > @Alessio-Parmeggiani can you pls rebase this commit? We are very close to finish it. Thanks
> 
> Sorry, what do you mean? I am not an expert in this kind of git stuff

It is required to solve the conflicts below

**Comment by avisconti** [comment]
You have to rebase, not merge

File: drivers/sensor/lsm6dsl/lsm6dsl.c (hunk 4 above)
Comment (fabiobaltieri) [reply drivers/sensor/lsm6dsl/lsm6dsl.c]: I can't recall a specific guideline, it's generally discouraged depending on the reviewer (I would flag it) unless it really improves the code readability.

Blocks or declare and initialize at the top are commonly used (all the cfg/data pairs for example), in this case I'd say either declare and initialize both `dval` and `dval_ug` or declare and initialize separately both, keep the styling coherent with the function and ideally with the whole file.

Comment (fabiobaltieri) [reply drivers/sensor/lsm6dsl/lsm6dsl.c]: Also the comment format is off, but also the comment are redundant, the variable name already includes the unit (as it should be), drop the comment

**Comment by fabiobaltieri** [comment]
Needs a rebase, also a commit message description and sign-off-by, see https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow (or ask for help on discord)

**Comment by avisconti** [comment]
replaced  by #72367

**Pull request closed without merging**